Firestore/Swift/Tests/Codable/CodableDocumentTests.swift

```diff
@@ -0,0 +1,550 @@
+/*
+ * Copyright 2019 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import Foundation
+import FirebaseFirestore
+@testable import FirebaseFirestoreSwift
+import XCTest
+
+private func assertRoundTrip<X: Equatable & Codable>(model: X, encoded: [String: Any]) -> Void {
+  let enc = assertEncodes(model, encoded: encoded)
+  assertDecodes(enc, encoded: model)
+}
+
+private func assertEncodes<X: Equatable & Codable>(_ model: X, encoded: [String: Any]) -> [String: Any] {
+  do {
+    let enc = try Firestore.Encoder().encode(model)
+    XCTAssertEqual(enc as NSDictionary, encoded as NSDictionary)
+    return enc
+  } catch {
+    XCTFail("Failed to encode \(X.self): error: \(error)")
+  }
+  return ["": -1]
+}
+
+private func assertDecodes<X: Equatable & Codable>(_ model: [String: Any], encoded: X) -> Void {
+  do {
+    let decoded = try Firestore.Decoder().decode(X.self, from: model)
+    XCTAssertEqual(decoded, encoded)
+  } catch {
+    XCTFail("Failed to decode \(X.self): \(error)")
+  }
+}
+
+private func assertEncodingThrows<X: Equatable & Codable>(_ model: X) -> Void {
+  do {
+    _ = try Firestore.Encoder().encode(model)
+  } catch {
+    return
+  }
+  XCTFail("Failed to throw")
+}
+
+private func assertDecodingThrows<X: Equatable & Codable>(_ model: [String: Any], encoded: X) -> Void {
```

Comment:
`encoded` appears unused. Remove it?

Reply:
compiler complains about X not used in function signature after removing.